---
title: "Home"
author: "Costa, W. G."
date: "`r Sys.Date()`"
site: workflowr::wflow_site
url: https://wevertongomescosta.github.io/Genomic-prediction-through-machine-learning-and-neural-networks-for-traits-with-epistasis/
output:
  workflowr::wflow_html:
    toc: yes
  highlight: github
editor_options:
  chunk_output_type: console
github-repo: wevertongomescosta/Genomic-prediction-through-machine-learning-and-neural-networks-for-traits-with-epistasis
---

# Genomic prediction through machine learning and neural networks for traits with epistasis

# Creating Genetic Map Chart

## Libraries 
```{r}
library(tidyverse)
library(ggrepel)
library(ggpubr)
library(cowplot)
```

## Genetic map

Primeiro vamos definir nossos SNPs de interesse para as variáveis. Essas informações foram pré-definidas e podem ser encontrados no arquivo `control_genetic`![](data/simulated_data/control_genetic.txt)

```{r}
snpsOfInterest <- data.frame(
  rbind(
    cbind(
      rep(1, 8), c(201,	602,	1003,	1404,	1805,	2206,	2607,	3008)),
    cbind(
      rep(2, 40),
      c(41,
      121,
      201,
      281,
      361,
      442,
      522,
      602,
      682,
      762,
      843,
      923,
      1003,
      1083,
      1163,
      1244,
      1324,
      1404,
      1484,
      1564,
      1645,
      1725,
      1805,
      1885,
      1965,
      2046,
      2126,
      2206,
      2286,
      2366,
      2447,
      2527,
      2607,
      2687,
      2767,
      2848,
      2928,
      3008,
      3088,
      3168
    )
  ),
  cbind(
    rep(3, 80),
    c(
      21,
      61,
      101,
      141,
      181,
      221,
      261,
      301,
      341,
      381,
      422,
      462,
      502,
      542,
      582,
      622,
      662,
      702,
      742,
      782,
      823,
      863,
      903,
      943,
      983,
      1023,
      1063,
      1103,
      1143,
      1183,
      1224,
      1264,
      1304,
      1344,
      1384,
      1424,
      1464,
      1504,
      1544,
      1584,
      1625,
      1665,
      1705,
      1745,
      1785,
      1825,
      1865,
      1905,
      1945,
      1985,
      2026,
      2066,
      2106,
      2146,
      2186,
      2226,
      2266,
      2306,
      2346,
      2386,
      2427,
      2467,
      2507,
      2547,
      2587,
      2627,
      2667,
      2707,
      2747,
      2787,
      2828,
      2868,
      2908,
      2948,
      2988,
      3028,
      3068,
      3108,
      3148,
      3188
    )
  ),
  cbind(
    rep(4, 120),
    c(
      25,
      50,
      75,
      100,
      125,
      150,
      175,
      200,
      225,
      250,
      275,
      300,
      325,
      350,
      375,
      426,
      451,
      476,
      501,
      526,
      551,
      576,
      601,
      626,
      651,
      676,
      701,
      726,
      751,
      776,
      827,
      852,
      877,
      902,
      927,
      952,
      977,
      1002,
      1027,
      1052,
      1077,
      1102,
      1127,
      1152,
      1177,
      1228,
      1253,
      1278,
      1303,
      1328,
      1353,
      1378,
      1403,
      1428,
      1453,
      1478,
      1503,
      1528,
      1553,
      1578,
      1629,
      1654,
      1679,
      1704,
      1729,
      1754,
      1779,
      1804,
      1829,
      1854,
      1879,
      1904,
      1929,
      1954,
      1979,
      2030,
      2055,
      2080,
      2105,
      2130,
      2155,
      2180,
      2205,
      2230,
      2255,
      2280,
      2305,
      2330,
      2355,
      2380,
      2431,
      2456,
      2481,
      2506,
      2531,
      2556,
      2581,
      2606,
      2631,
      2656,
      2681,
      2706,
      2731,
      2756,
      2781,
      2832,
      2857,
      2882,
      2907,
      2932,
      2957,
      2982,
      3007,
      3032,
      3057,
      3082,
      3107,
      3132,
      3157,
      3182
    )
  ),
  cbind(
    rep(5, 240),
    c(
      11,
      24,
      37,
      50,
      63,
      76,
      89,
      102,
      115,
      128,
      141,
      154,
      167,
      180,
      193,
      206,
      219,
      232,
      245,
      258,
      271,
      284,
      297,
      310,
      323,
      336,
      349,
      362,
      375,
      388,
      412,
      425,
      438,
      451,
      464,
      477,
      490,
      503,
      516,
      529,
      542,
      555,
      568,
      581,
      594,
      607,
      620,
      633,
      646,
      659,
      672,
      685,
      698,
      711,
      724,
      737,
      750,
      763,
      776,
      789,
      813,
      826,
      839,
      852,
      865,
      878,
      891,
      904,
      917,
      930,
      943,
      956,
      969,
      982,
      995,
      1008,
      1021,
      1034,
      1047,
      1060,
      1073,
      1086,
      1099,
      1112,
      1125,
      1138,
      1151,
      1164,
      1177,
      1190,
      1214,
      1227,
      1240,
      1253,
      1266,
      1279,
      1292,
      1305,
      1318,
      1331,
      1344,
      1357,
      1370,
      1383,
      1396,
      1409,
      1422,
      1435,
      1448,
      1461,
      1474,
      1487,
      1500,
      1513,
      1526,
      1539,
      1552,
      1565,
      1578,
      1591,
      1615,
      1628,
      1641,
      1654,
      1667,
      1680,
      1693,
      1706,
      1719,
      1732,
      1745,
      1758,
      1771,
      1784,
      1797,
      1810,
      1823,
      1836,
      1849,
      1862,
      1875,
      1888,
      1901,
      1914,
      1927,
      1940,
      1953,
      1966,
      1979,
      1992,
      2016,
      2029,
      2042,
      2055,
      2068,
      2081,
      2094,
      2107,
      2120,
      2133,
      2146,
      2159,
      2172,
      2185,
      2198,
      2211,
      2224,
      2237,
      2250,
      2263,
      2276,
      2289,
      2302,
      2315,
      2328,
      2341,
      2354,
      2367,
      2380,
      2393,
      2417,
      2430,
      2443,
      2456,
      2469,
      2482,
      2495,
      2508,
      2521,
      2534,
      2547,
      2560,
      2573,
      2586,
      2599,
      2612,
      2625,
      2638,
      2651,
      2664,
      2677,
      2690,
      2703,
      2716,
      2729,
      2742,
      2755,
      2768,
      2781,
      2794,
      2818,
      2831,
      2844,
      2857,
      2870,
      2883,
      2896,
      2909,
      2922,
      2935,
      2948,
      2961,
      2974,
      2987,
      3000,
      3013,
      3026,
      3039,
      3052,
      3065,
      3078,
      3091,
      3104,
      3117,
      3130,
      3143,
      3156,
      3169,
      3182,
      3195
    )
  ),
  cbind(
    rep(6, 480),
    c(
      23,
      29,
      35,
      41,
      47,
      53,
      59,
      65,
      71,
      77,
      83,
      89,
      95,
      101,
      107,
      113,
      119,
      125,
      131,
      137,
      143,
      149,
      155,
      161,
      167,
      173,
      179,
      185,
      191,
      197,
      203,
      209,
      215,
      221,
      227,
      233,
      239,
      245,
      251,
      257,
      263,
      269,
      275,
      281,
      287,
      293,
      299,
      305,
      311,
      317,
      323,
      329,
      335,
      341,
      347,
      353,
      359,
      365,
      371,
      377,
      424,
      430,
      436,
      442,
      448,
      454,
      460,
      466,
      472,
      478,
      484,
      490,
      496,
      502,
      508,
      514,
      520,
      526,
      532,
      538,
      544,
      550,
      556,
      562,
      568,
      574,
      580,
      586,
      592,
      598,
      604,
      610,
      616,
      622,
      628,
      634,
      640,
      646,
      652,
      658,
      664,
      670,
      676,
      682,
      688,
      694,
      700,
      706,
      712,
      718,
      724,
      730,
      736,
      742,
      748,
      754,
      760,
      766,
      772,
      778,
      825,
      831,
      837,
      843,
      849,
      855,
      861,
      867,
      873,
      879,
      885,
      891,
      897,
      903,
      909,
      915,
      921,
      927,
      933,
      939,
      945,
      951,
      957,
      963,
      969,
      975,
      981,
      987,
      993,
      999,
      1005,
      1011,
      1017,
      1023,
      1029,
      1035,
      1041,
      1047,
      1053,
      1059,
      1065,
      1071,
      1077,
      1083,
      1089,
      1095,
      1101,
      1107,
      1113,
      1119,
      1125,
      1131,
      1137,
      1143,
      1149,
      1155,
      1161,
      1167,
      1173,
      1179,
      1226,
      1232,
      1238,
      1244,
      1250,
      1256,
      1262,
      1268,
      1274,
      1280,
      1286,
      1292,
      1298,
      1304,
      1310,
      1316,
      1322,
      1328,
      1334,
      1340,
      1346,
      1352,
      1358,
      1364,
      1370,
      1376,
      1382,
      1388,
      1394,
      1400,
      1406,
      1412,
      1418,
      1424,
      1430,
      1436,
      1442,
      1448,
      1454,
      1460,
      1466,
      1472,
      1478,
      1484,
      1490,
      1496,
      1502,
      1508,
      1514,
      1520,
      1526,
      1532,
      1538,
      1544,
      1550,
      1556,
      1562,
      1568,
      1574,
      1580,
      1627,
      1633,
      1639,
      1645,
      1651,
      1657,
      1663,
      1669,
      1675,
      1681,
      1687,
      1693,
      1699,
      1705,
      1711,
      1717,
      1723,
      1729,
      1735,
      1741,
      1747,
      1753,
      1759,
      1765,
      1771,
      1777,
      1783,
      1789,
      1795,
      1801,
      1807,
      1813,
      1819,
      1825,
      1831,
      1837,
      1843,
      1849,
      1855,
      1861,
      1867,
      1873,
      1879,
      1885,
      1891,
      1897,
      1903,
      1909,
      1915,
      1921,
      1927,
      1933,
      1939,
      1945,
      1951,
      1957,
      1963,
      1969,
      1975,
      1981,
      2028,
      2034,
      2040,
      2046,
      2052,
      2058,
      2064,
      2070,
      2076,
      2082,
      2088,
      2094,
      2100,
      2106,
      2112,
      2118,
      2124,
      2130,
      2136,
      2142,
      2148,
      2154,
      2160,
      2166,
      2172,
      2178,
      2184,
      2190,
      2196,
      2202,
      2208,
      2214,
      2220,
      2226,
      2232,
      2238,
      2244,
      2250,
      2256,
      2262,
      2268,
      2274,
      2280,
      2286,
      2292,
      2298,
      2304,
      2310,
      2316,
      2322,
      2328,
      2334,
      2340,
      2346,
      2352,
      2358,
      2364,
      2370,
      2376,
      2382,
      2429,
      2435,
      2441,
      2447,
      2453,
      2459,
      2465,
      2471,
      2477,
      2483,
      2489,
      2495,
      2501,
      2507,
      2513,
      2519,
      2525,
      2531,
      2537,
      2543,
      2549,
      2555,
      2561,
      2567,
      2573,
      2579,
      2585,
      2591,
      2597,
      2603,
      2609,
      2615,
      2621,
      2627,
      2633,
      2639,
      2645,
      2651,
      2657,
      2663,
      2669,
      2675,
      2681,
      2687,
      2693,
      2699,
      2705,
      2711,
      2717,
      2723,
      2729,
      2735,
      2741,
      2747,
      2753,
      2759,
      2765,
      2771,
      2777,
      2783,
      2830,
      2836,
      2842,
      2848,
      2854,
      2860,
      2866,
      2872,
      2878,
      2884,
      2890,
      2896,
      2902,
      2908,
      2914,
      2920,
      2926,
      2932,
      2938,
      2944,
      2950,
      2956,
      2962,
      2968,
      2974,
      2980,
      2986,
      2992,
      2998,
      3004,
      3010,
      3016,
      3022,
      3028,
      3034,
      3040,
      3046,
      3052,
      3058,
      3064,
      3070,
      3076,
      3082,
      3088,
      3094,
      3100,
      3106,
      3112,
      3118,
      3124,
      3130,
      3136,
      3142,
      3148,
      3154,
      3160,
      3166,
      3172,
      3178,
      3184
    )
  )
))
```

Agora vamos definir os nomes das colunas para nosso `snpsOfInterest` e criar um objeto `locus` com as informações de ininício e témino de cada grupo de ligação. 

```{r}
colnames(snpsOfInterest) <- c("variable", "marker")

locus <-
  data.frame(
    c(
      1,
      401,
      402,
      802,
      803,
      1203,
      1204,
      1604,
      1605,
      2005,
      2006,
      2406,
      2407,
      2807,
      2808,
      3208,
      3209,
      3609,
      3610,
      4010
    )
  )

colnames(locus) <- c("marker")
```

Para facilitar a visualização criei um gráfico do mapa genético das cartacterísticas. Para isso criei o  `map` com o número de marcadores e tamanho de cada grupo de liagação. 

```{r}
letters <- c("A", "B", "C", "D", "E")

map <-
  data.frame(rbind(
    cbind(seq(1, 401, 1), rep("LG 1", 401), seq(0, 200, 0.5)),
    cbind(seq(402, 802, 1), rep("LG 2", 401), seq(0, 200, 0.5)),
    cbind(seq(803, 1203, 1), rep("LG 3", 401), seq(0, 200, 0.5)),
    cbind(seq(1204, 1604, 1), rep("LG 4", 401), seq(0, 200, 0.5)),
    cbind(seq(1605, 2005, 1), rep("LG 5", 401), seq(0, 200, 0.5)),
    cbind(seq(2006, 2406, 1), rep("LG 6", 401), seq(0, 200, 0.5)),
    cbind(seq(2407, 2807, 1), rep("LG 7", 401), seq(0, 200, 0.5)),
    cbind(seq(2808, 3208, 1), rep("LG 8", 401), seq(0, 200, 0.5)),
    cbind(seq(3209, 3609, 1), rep("LG 9", 401), seq(0, 200, 0.5)),
    cbind(seq(3610, 4010, 1), rep("LG 10", 401), seq(0, 200, 0.5))
  ))

colnames(map) <- c("marker", "LG", "Size")

map <- map %>%
  mutate(
    marker = as.numeric(marker),
    Size = as.numeric(Size),
    LG = factor(
      LG,
      levels = c(
        "LG 1",
        "LG 2",
        "LG 3",
        "LG 4",
        "LG 5",
        "LG 6",
        "LG 7",
        "LG 8",
        "LG 9",
        "LG 10"
      )
    )
  )
```

Para dividir a figura e mostrar todos os mapas genômicos das características, dividi o `snpsOfInterest` e o `map` para cada característica e inclui os SNPs de interesse no `map` para cada característica.

```{r}
snpsOfInterest1 <- snpsOfInterest %>%
  filter(variable == 1)
snpsOfInterest2 <- snpsOfInterest %>%
  filter(variable == 2)
snpsOfInterest3 <- snpsOfInterest %>%
  filter(variable == 3)
snpsOfInterest4 <- snpsOfInterest %>%
  filter(variable == 4)
snpsOfInterest5 <- snpsOfInterest %>%
  filter(variable == 5)
snpsOfInterest6 <- snpsOfInterest %>%
  filter(variable == 6)

map1 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest1$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )
map2 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest2$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )
map3 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest3$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )
map4 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest4$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )
map5 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest5$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )
map6 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest6$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )
```

Agora cirei o gráfico de cada característica e depois agrupei eles em apenas uma imagem `maps`.

```{r}
mapa1 <- ggplot(map1, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map1, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map1, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map1, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map1, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

mapa2 <- ggplot(map2, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map2, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map2, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map2, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map2, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

mapa3 <- ggplot(map3, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map3, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map3, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map3, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map3, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

mapa4 <- ggplot(map4, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map4, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map4, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map4, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map4, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

mapa5 <- ggplot(map5, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map5, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map5, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map5, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map5, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

mapa6 <- ggplot(map6, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map6, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map6, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map6, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map6, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")


maps <- ggdraw() +
  draw_plot(
    mapa1,
    x = 0.05,
    y = .5,
    width = .3,
    height = .5
  ) +
  draw_plot(
    mapa2,
    x = .4,
    y = .5,
    width = .3,
    height = .5
  ) +
  draw_plot(
    mapa3,
    x = .75,
    y = .5,
    width = .3,
    height = .5
  ) +
  draw_plot(
    mapa4,
    x = 0.25,
    y = 0,
    width = 0.3,
    height = 0.5
  ) +
  draw_plot(
    mapa5,
    x = 0.65,
    y = 0,
    width = 0.3,
    height = 0.5
  ) +
  draw_plot_label(
    label = c("A", "B", "C", "D", "E"),
    size = 15,
    x = c(0, 0.35, 0.7, 0.2, 0.6),
    y = c(1, 1, 1, 0.5, 0.5)
  )

print(maps)

#ggsave(filename="output/map.tiff", plot = maps, units = "cm", width = 17.5, height = 17.5, dpi = 300)
```

# Data

## Data importantion

```{r}
pathFenT <- file.path("data/PHEN_T")
pathFenV <- file.path("data/PHEN_V")
pathGenT <- file.path("data/GEN_T")
pathGenV <- file.path("data/GEN_V")
```

## Listen files cross-validation

```{r}
filesFenT <-
  list.files(path = pathFenT,
             pattern = "/*.txt",
             full.names = T)
filesFenV <-
  list.files(path = pathFenV,
             pattern = "/*.txt",
             full.names = T)
filesGenT <-
  list.files(path = pathGenT,
             pattern = "/*.txt",
             full.names = T)
filesGenV <-
  list.files(path = pathGenV,
             pattern = "/*.txt",
             full.names = T)
```

## Defining numbers of variabels and folds

```{r}
nvariable <- 18
nfolds <- 5
```

# Analysis

```{r}
set.seed(123) #Seed for reproduction of results
```

## Multivariate Adaptive Regression Splines (MARS)

### Libraries 

```{r}
library(earth)                # fit MARS models
library(caret); library(vip)  # Variable importance
library(tidyverse)            # Data Handling
```

### Linear MARS Processing
   
```{r}
for(j in 1:nvariable) {
  cat("===========================================================",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("===========================================================",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1) #Objeto RQEM training
  reqv <- matrix(nrow = nfolds, ncol = 1) #RQEM validation object
  
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    
    # **************************************************** ******************
    # Fit a basic MARS model LINEAR - Training
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    mars1 <- earth(x = dataTx, y = dataTy[, j])
    coefl <- mars1$coefficients
    
    ## Model parameters
    rt = cor(mars1$fitted.values, dataTy[, j])
    R2t[i] <- rt * rt
    errort = mars1$fitted.values - dataTy[, j]
    reqt[i] = sqrt(mean(errort ^ 2))
    
    ## Importance of bookmarks
    
    imp <- evimp(mars1, trim = FALSE)
    imp <- as.data.frame(unclass(imp[, c(1, 6)]))
    names <- cbind(imp$col, i)
    imp <- data.frame(imp$rss, names)
    colnames(imp) <- c("Overall", "marker", "n fold")
    
    if (i == 1) {
      imp.mars1 <- imp
    } else {
      imp.mars1 <- imp.mars1 %>% rbind(imp)
    }
    
    ## Coefficients
    #print(coefl)
    
    # **************************************************** *****************
    # Validation
    # **************************************************** *****************
    
    ## Linear model prediction
    ypred <- predict(mars1, dataVx)
    
    ## Model parameters
    rv = cor(ypred, dataVy[, j])
    R2v[i] = rv * rv
    errorv = dataVy[, j] - ypred
    reqv[i] = sqrt(mean(errorv ^ 2))
  }
  
  cat("Variable model parameters", j, "\n")
  par.mars1 <- cbind(R2t, R2v, reqt, reqv)
  par.mars1 <-
    rbind(par.mars1, apply(par.mars1, 2, mean), apply(par.mars1, 2, sd))
  colnames(par.mars1) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.mars1) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.mars1), "MARS L", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.mars1 <- data.frame(par.mars1, names)
  par.mars1
  
  # Result of all variables
  
  if (j == 1) {
    res.mars1 <- par.mars1
  } else {
    res.mars1 <- res.mars1 %>% rbind(par.mars1)
  }
  
  # Importance of all variable markers
  
  names <- cbind("MARS L", rep(j, len = nrow(imp.mars1)))
  colnames(names) <- c("method", "variable")
  imp.mars1 <- data.frame(imp.mars1, names)
  
  if (j == 1) {
    res.imp.mars1 <- imp.mars1
  } else {
    res.imp.mars1 <- res.imp.mars1 %>% rbind(imp.mars1)
  }
}

cat("Final result of all variables for MARS Linear", "\n")
print(res.mars1)

cat("Importance of markers for all MARS Linear variables", "\n")
res.imp.mars1
```


### Quadratic MARS Processing

```{r}
for (j in 1:nvariable) {
  cat("==================================================== =======",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("==================================================== =======",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1)# Training RQEM object
  reqv <- matrix(nrow = nfolds, ncol = 1)#RQEM validation object
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    # **************************************************** ******************
    # Fit a basic MARS model SQUARE
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    mars2 <- earth(x = dataTx, y = dataTy[, j], degree = 2)
    coefq <- list(mars2$coefficients)
    
    ## Model parameters
    rt = cor(mars2$fitted.values, dataTy[, j])
    R2t[i] = rt * rt
    errort = mars2$fitted.values - dataTy[, j]
    reqt[i] = sqrt(mean(errort ^ 2))
    
    ## Importance of bookmarks
    imp <- evimp(mars2, trim = FALSE)
    imp <- as.data.frame(unclass(imp[, c(1, 6)]))
    names <- cbind(imp$col, i)
    imp <- data.frame(imp$rss, names)
    colnames(imp) <- c("Overall", "marker", "n fold")
    
    if (i == 1) {
      imp.mars2 <- imp
    } else {
      imp.mars2 <- imp.mars2 %>% rbind(imp)
    }
    
    
    ## Coefficients
    # print(coefq)
    
    # **************************************************** ******************
    # 6. Validation
    # **************************************************** ******************
    
    ## Quadratic model prediction
    ypred <- predict(mars2, dataVx)
    
    
    ##Model Parameters
    rv = cor(ypred, dataVy[, j])
    R2v[i] = rv * rv
    errorv = dataVy[, j] - ypred
    reqv[i] = sqrt(mean(errorv ^ 2))
  }
  
  cat("Variable model parameters", j, "\n")
  par.mars2 <- cbind(R2t, R2v, reqt, reqv)
  par.mars2 <-
    rbind(par.mars2, apply(par.mars2, 2, mean), apply(par.mars2, 2, sd))
  colnames(par.mars2) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.mars2) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.mars2), "MARS Q", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.mars2 <- data.frame(par.mars2, names)
  par.mars2
  
  # Result of all variables
  if (j == 1) {
    res.mars2 <- par.mars2
  } else {
    res.mars2 <- res.mars2 %>% rbind(par.mars2)
  }
  
  #Importance of all variable markers
  names <- cbind("MARS Q", rep(j, len = nrow(imp.mars2)))
  colnames(names) <- c("method", "variable")
  imp.mars2 <- data.frame(imp.mars2, names)
  
  if (j == 1) {
    res.imp.mars2 <- imp.mars2
  } else {
    res.imp.mars2 <- res.imp.mars2 %>% rbind(imp.mars2)
  }
}

cat("Final result of all variables for Quadratic MARS", "\n")
res.mars2

cat("Importance of markers for all Quadratic MARS variables", "\n")
res.imp.mars2

```

### Cubic MARS Processing

```{r}
for (j in 1:nvariable) {
  cat("==================================================== =======",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("==================================================== =======",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1)# Training RQEM object
  reqv <- matrix(nrow = nfolds, ncol = 1)#RQEM validation object
  
  
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    # **************************************************** ******************
    # Fit a basic MARS model Cubico
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    mars3 <-
      earth(
        x = dataTx,
        y = dataTy[, j],
        degree = 3,
        thresh = 0.005
      ) #Change in R² less than 0.05, stopping criterion
    coefc <- list(mars3$coefficients)
    
    ## Model parameters
    rt = cor(mars3$fitted.values, dataTy[, j])
    R2t[i] = rt * rt
    errort = mars3$fitted.values - dataTy[, j]
    reqt[i] = sqrt(mean(errort ^ 2))
    
    ## Importance of bookmarks
    imp <- evimp(mars3, trim = FALSE)
    imp <- as.data.frame(unclass(imp[, c(1, 6)]))
    names <- cbind(imp$col, i)
    imp <- data.frame(imp$rss, names)
    colnames(imp) <- c("Overall", "marker", "n fold")
    
    if (i == 1) {
      imp.mars3 <- imp
    } else {
      imp.mars3 <- imp.mars3 %>% rbind(imp)
    }
    
    
    ## Coefficients
    # print(coefc)
    
    # **************************************************** ******************
    # 6. Validation
    # **************************************************** ******************
    
    ## Cubic model prediction
    ypred <- predict(mars3, dataVx)
    
    
    ##Model Parameters
    rv = cor(ypred, dataVy[, j])
    R2v[i] = rv * rv
    errorv = dataVy[, j] - ypred
    reqv[i] = sqrt(mean(errorv ^ 2))
  }
  
  cat("Variable model parameters", j, "\n")
  par.mars3 <- cbind(R2t, R2v, reqt, reqv)
  par.mars3 <-
    rbind(par.mars3, apply(par.mars3, 2, mean), apply(par.mars3, 2, sd))
  colnames(par.mars3) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.mars3) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.mars3), "MARS C", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.mars3 <- data.frame(par.mars3, names)
  par.mars3
  
  # Result of all variables
  if (j == 1) {
    res.mars3 <- par.mars3
  } else {
    res.mars3 <- res.mars3 %>% rbind(par.mars3)
  }
  
  #Importance of all variable markers
  names <- cbind("MARS C", rep(j, len = nrow(imp.mars3)))
  colnames(names) <- c("method", "variable")
  imp.mars3 <- data.frame(imp.mars3, names)
  
  if (j == 1) {
    res.imp.mars3 <- imp.mars3
  } else {
    res.imp.mars3 <- res.imp.mars3 %>% rbind(imp.mars3)
  }
}

cat("Final result of all variables for Cubic MARS", "\n")
res.mars3

cat("Importance of markers for all Cubic MARS variables", "\n")
res.imp.mars3
```

## Decision Tree - Continuous Response
### Load packages

```{r}
library(rpart)
library(ISLR)
```

### Decision Tree Processing

```{r}
for(j in 1:nvariable) {
  cat("==================================================== =======",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("==================================================== =======",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1)# Training RQEM object
  reqv <- matrix(nrow = nfolds, ncol = 1)#RQEM validation object
  
  
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    
    # **************************************************** ******************
    # Fit a basic Decision Tree - Training
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    fit_tree <- rpart(dataTy[, j] ~ . , data = dataTx)
    
    ## Model parameters
    rs = cor(dataTy[, j], predict(fit_tree))
    R2t[i] = rs * rs
    errort = predict(fit_tree) - dataTy[, j]
    reqt[i] = sqrt(mean(errort ^ 2))
    
    ## Importance of bookmarks
    imp <-
      data.frame(varImp(fit_tree, scale = TRUE, value = "rss"))
    
    names <- cbind(rownames(imp), i)
    colnames(names) <- c("marker", "n fold")
    imp <- data.frame(imp, names)
    
    if (i == 1) {
      imp.ad <- imp
    } else {
      imp.ad <- imp.ad %>% rbind(imp)
    }
    
    
    # **************************************************** ******************
    # Validation
    # **************************************************** ******************
    
    ## Model prediction
    outputV = predict(fit_tree, newdata = dataVx)
    
    
    ## Model parameters
    rs = cor(dataVy[, j], outputV)
    R2v[i] = rs * rs
    errorv = dataVy[, j] - outputV
    reqv[i] = sqrt(mean(errorv ^ 2))
  }
  
  cat("Variable model parameters", j, "\n")
  par.ad <- cbind(R2t, R2v, reqt, reqv)
  par.ad <- rbind(par.ad, apply(par.ad, 2, mean), apply(par.ad, 2, sd))
  colnames(par.ad) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.ad) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.ad), "DT", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.ad <- data.frame(par.ad, names)
  par.ad
  
  # Result of all variables
  if (j == 1) {
    res.ad <- par.ad
  } else {
    res.ad <- res.ad %>% rbind(par.ad)
  }
  
  #Importance of all variable markers
  names <- cbind("DT", rep(j, len = nrow(imp.ad)))
  colnames(names) <- c("method", "variable")
  imp.ad <- data.frame(imp.ad, names)
  
  
  if (j == 1) {
    res.imp.ad <- imp.ad
  } else {
    res.imp.ad <- res.imp.ad %>% rbind(imp.ad)
  }
}
cat("Final result of all variables for Decision Tree", "\n")
res.ad

cat("Importance of markers for all decision tree variables", "\n")
res.imp.ad
```

## Bagging

### Load packages

```{r}
library(randomForest)
```

#2. Bagging Processing

```{r}
for(j in 1:nvariable) {
  cat("==================================================== =======",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("==================================================== =======",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1)# Training RQEM object
  reqv <- matrix(nrow = nfolds, ncol = 1)#RQEM validation object
  
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    
    # **************************************************** ******************
    # Fit a basic Bagging - Training
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    bag <-
      randomForest(dataTy[, j] ~ . , data = dataTx, mtry = ncol(dataTx))
    
    ## Model parameters
    rs = cor(dataTy[, j], predict(bag))
    R2t[i] = rs * rs
    errort = predict(bag) - dataTy[, j]
    reqt[i] = sqrt(mean(errort ^ 2))
    
    ## Importance of bookmarks
    imp <- data.frame(varImp(bag, scale = TRUE, value = "rss"))
    
    names <- cbind(rownames(imp), i)
    colnames(names) <- c("marker", "n fold")
    imp <- data.frame(imp, names)
    
    if (i == 1) {
      imp.bag <- imp
    } else {
      imp.bag <- imp.bag %>% rbind(imp)
    }
    
    
    # **************************************************** ******************
    # Validation
    # **************************************************** ******************
    
    ## Model prediction
    outputV = predict(bag, newdata = dataVx)
    
    
    ## Model parameters
    rs = cor(dataVy[, j], outputV)
    R2v[i] = rs * rs
    errorv = dataVy[, j] - outputV
    reqv[i] = sqrt(mean(errorv ^ 2))
  }
  
  cat("Variable model parameters", j, "\n")
  par.bag <- cbind(R2t, R2v, reqt, reqv)
  par.bag <-
    rbind(par.bag, apply(par.bag, 2, mean), apply(par.bag, 2, sd))
  colnames(par.bag) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.bag) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.bag), "BAG", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.bag <- data.frame(par.bag, names)
  par.bag
  
  # Result of all variables
  if (j == 1) {
    res.bag <- par.bag
  } else {
    res.bag <- res.bag %>% rbind(par.bag)
  }
  
  #Importance of all variable markers
  names <- cbind("BAG", rep(j, len = nrow(imp.bag)))
  colnames(names) <- c("method", "variable")
  imp.bag <- data.frame(imp.bag, names)
  
  if (j == 1) {
    res.imp.bag <- imp.bag
  } else {
    res.imp.bag <- res.imp.bag %>% rbind(imp.bag)
  }
}
cat("Final result of all variables for Bagging", "\n")
res.bag

cat("Importance of markers for all Bagging variables", "\n")
res.imp.bag
```

## Random Forest

### Loading packages

```{r}
library(randomForest)
```

#2. Random Forest Processing

```{r}
for(j in 1:nvariable) {
  cat("==================================================== =======",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("==================================================== =======",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1)# Training RQEM object
  reqv <- matrix(nrow = nfolds, ncol = 1)#RQEM validation object
  
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    
    # **************************************************** ******************
    # Fit a basic Random Forest - Training
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    rf <-
      randomForest(
        dataTy[, j] ~ . ,
        mytry = (ncol(dataTx) / 3),
        data = dataTx,
        ntree = 500
      )
    
    ## Model parameters
    rs = cor(dataTy[, j], predict(rf))
    R2t[i] = rs * rs
    errort = predict(rf) - dataTy[, j]
    reqt[i] = sqrt(mean(errort ^ 2))
    
    ## Importance of bookmarks
    imp <- data.frame(varImp(rf, scale = TRUE, value = "rss"))
    
    names <- cbind(rownames(imp), i)
    colnames(names) <- c("marker", "n fold")
    imp <- data.frame(imp, names)
    
    if (i == 1) {
      imp.rf <- imp
    } else {
      imp.rf <- imp.rf %>% rbind(imp)
    }
    
    # **************************************************** ******************
    # Validation
    # **************************************************** ******************
    
    ## Model prediction
    outputV = predict(rf, newdata = dataVx)
    
    
    ## Model parameters
    rs = cor(dataVy[, j], outputV)
    R2v[i] = rs * rs
    errorv = dataVy[, j] - outputV
    reqv[i] = sqrt(mean(errorv ^ 2))
  }
  
  cat("Variable model parameters", j, "\n")
  par.rf <- cbind(R2t, R2v, reqt, reqv)
  par.rf <- rbind(par.rf, apply(par.rf, 2, mean), apply(par.rf, 2, sd))
  colnames(par.rf) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.rf) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.rf), "RF", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.rf <- data.frame(par.rf, names)
  par.rf
  
  # Result of all variables
  if (j == 1) {
    res.rf <- par.rf
  } else {
    res.rf <- res.rf %>% rbind(par.rf)
  }
  
  #Importance of all variable markers
  names <- cbind("RF", rep(j, len = nrow(imp.rf)))
  colnames(names) <- c("method", "variable")
  imp.rf <- data.frame(imp.rf, names)
  
  if (j == 1) {
    res.imp.rf <- imp.rf
  } else {
    res.imp.rf <- res.imp.rf %>% rbind(imp.rf)
  }
}
cat("Final result of all variables for Random Forest", "\n")
res.rf

cat("Importance of markers for all variables in Random Forest", "\n")
res.imp.rf
```

## Boosting

### Load packages

```{r}
library(gbm)
```

### Boosting Processing

```{r}
for(j in 1:nvariable) {
  cat("==================================================== =======",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("==================================================== =======",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1)# Training RQEM object
  reqv <- matrix(nrow = nfolds, ncol = 1)#RQEM validation object
  
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    
    # **************************************************** ******************
    # Fit a basic Boosting - Training
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    boost = gbm(
      dataTy[, j] ~ . ,
      data = dataTx,
      distribution = "gaussian",
      n.trees = 500 ,
      interaction.depth = 2
    )
    
    ## Model parameters
    output1 = predict(boost, dataTx, n.trees = 500)
    rs = cor(dataTy[, j], output1)
    R2t[i] = rs * rs
    errort = output1 - dataVy[, j]
    reqt[i] = sqrt(mean(errort ^ 2))
    
    ## Importance of bookmarks
    imp <-
      data.frame(varImp(
        boost,
        scale = TRUE,
        numTrees = 500,
        value = "rss"
      ))
    
    names <- cbind(rownames(imp), i)
    colnames(names) <- c("marker", "n fold")
    imp <- data.frame(imp, names)
    
    if (i == 1) {
      imp.boost <- imp
    } else {
      imp.boost <- imp.boost %>% rbind(imp)
    }
    
    # **************************************************** ******************
    # Validation
    # **************************************************** ******************
    
    ## Model prediction
    outputV = predict(boost, newdata = dataVx, n.trees = 500)
    
    
    ## Model parameters
    rs = cor(dataVy[, j], outputV)
    R2v[i] = rs * rs
    errorv = outputV - dataVy[, j]
    reqv[i] = sqrt(mean(errorv ^ 2))
  }
  
  cat("Variable model parameters", j, "\n")
  par.boost <- cbind(R2t, R2v, reqt, reqv)
  par.boost <-
    rbind(par.boost, apply(par.boost, 2, mean), apply(par.boost, 2, sd))
  colnames(par.boost) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.boost) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.boost), "Boost", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.boost <- data.frame(par.boost, names)
  par.boost
  
  # Result of all variables
  if (j == 1) {
    res.boost <- par.boost
  } else {
    res.boost <- res.boost %>% rbind(par.boost)
  }
  
  #Importance of all variable markers
  names <- cbind("Boost", rep(j, len = nrow(imp.boost)))
  colnames(names) <- c("method", "variable")
  imp.boost <- data.frame(imp.boost, names)
  
  if (j == 1) {
    res.imp.boost <- imp.boost
  } else {
    res.imp.boost <- res.imp.boost %>% rbind(imp.boost)
  }
}
cat("Final result of all variables for MARS Linear", "\n")
res.boost

cat("Importance of markers for all MARS Linear variables", "\n")
res.imp.boost
```

## Genomic best linear unbiased prediction (G-BLUP)

### Loading packages

```{r}
library(rrBLUP)
library(MASS)
```

### G-BLUP Processing

```{r}
for(j in 1:nvariable) {
  cat("==================================================== =======",
      "\n")
  cat("Results for for variable ", j, "\n")
  cat("==================================================== =======",
      "\n")
  R2t <- matrix(nrow = nfolds, ncol = 1) #Object r² training
  R2v <- matrix(nrow = nfolds, ncol = 1) #Object r² validation
  reqt <- matrix(nrow = nfolds, ncol = 1)# Training RQEM object
  reqv <- matrix(nrow = nfolds, ncol = 1)#RQEM validation object
  
  for (i in 1:nfolds) {
    dataTy <- read.table(filesFenT[i])
    dataTx <- read.table(filesGenT[i])
    dataVy <- read.table(filesFenV[i])
    dataVx <- read.table(filesGenV[i])
    
    # **************************************************** ******************
    # Fit a G-LUP model - Training
    # **************************************************** ******************
    cat("K-fold = ", i, "\n")
    gblup = kinship.BLUP(dataTy[, j], G.train = dataTx)
    
    ## Model parameters
    rs = cor(dataTy[, j], gblup$g.train)
    R2t[i] = rs * rs
    errort <- dataTy[, j] - gblup$g.train
    errort = errort - mean(errort)
    reqt[i] <- sqrt(mean(errort ^ 2))
    
    
    ## Importance of bookmarks
    # structuring genomic values ​​in array to use sort
    gbv = as.matrix(gblup$g.train)
    
    # Marker effect vector
    dataTx = as.matrix(dataTx)
    effect.markers = ginv(t(dataTx) %*% dataTx) %*% (t(dataTx) %*% gbv)
    imp = as.data.frame(effect.markers)
    colnames(imp) <- "Overall"
    names <- cbind(colnames(dataTx), i)
    colnames(names) <- c("marker", "n fold")
    imp <- data.frame(imp, names)
    
    if (i == 1) {
      imp.gblup <- imp
    } else {
      imp.gblup <- imp.gblup %>% rbind(imp)
    }
    
    
    # **************************************************** ******************
    # Validation
    # **************************************************** ******************
    
    ## Model prediction
    dataVx = as.matrix(dataVx)
    outputV = dataVx %*% effect.markers
    
    ## Model parameters
    rs = cor(dataVy[, j], outputV)
    R2v[i] = rs * rs
    errorv <- dataVy[, j] - outputV
    errorv = errorv - mean(errorv)
    reqv[i] <- sqrt(mean(errorv ^ 2))
    
    
  }
  
  cat("Variable model parameters", j, "\n")
  par.gblup <- cbind(R2t, R2v, reqt, reqv)
  par.gblup <-
    rbind(par.gblup, apply(par.gblup, 2, mean), apply(par.gblup, 2, sd))
  colnames(par.gblup) <-
    c("R² Trein", "R² Val", "REQM Trein", "REQM Val")
  rownames(par.gblup) <-
    c("K-Fold 1",
      "K-Fold 2",
      "K-Fold 3",
      "K-Fold 4",
      "K-Fold 5",
      "Mean",
      " SD")
  names <- cbind(rownames(par.gblup), "G-BLUP", j)
  colnames(names) <- c("n Fold", "method", "variable")
  par.gblup <- data.frame(par.gblup, names)
  par.gblup
  
  # Result of all variables
  if (j == 1) {
    res.gblup <- par.gblup
  } else {
    res.gblup <- res.gblup %>% rbind(par.gblup)
  }
  
  #Importance of all variable markers
  names <- cbind("G-BLUP", rep(j, len = nrow(imp.gblup)))
  colnames(names) <- c("method", "variable")
  imp.gblup <- data.frame(imp.gblup, names)
  
  
  if (j == 1) {
    res.imp.gblup <- imp.gblup
  } else {
    res.imp.gblup <- res.imp.gblup %>% rbind(imp.gblup)
  }
}

cat("Final result of all variables for G-BLUP", "\n")
res.gblup

cat("Importance of markers for all G_BLUP variables", "\n")
res.imp.gblup
```

## Radial Base Network

### Loading results obtained in the [Genes Software](https://www.scielo.br/j/asagr/a/sLvDYF5MYv9kWR5MKgxb6sL/?lang=en). Scripts can be seen [here](analysis/suplementary__matlab_rbf.m)

```{r}
res.rbf <- read_excel("output/resultrbf.xls")

colnames(res.rbf) <- colnames(res.ad) # naming the columns like the other files
```
## Multilayer Perceptron Network

### Loading results obtained in the [Genes Software](https://www.scielo.br/j/asagr/a/sLvDYF5MYv9kWR5MKgxb6sL/?lang=en). Scripts can be seen [here](analysis/suplementary__matlab_MLP.m)

```{r}
res.mlp <- read_excel("output/resultmlp.xls")

colnames(res.mlp) <- colnames(res.ad) # naming the columns like the other files
```

## Results

### Organizing results

```{r}
result <-
  rbind(
    res.mars1,
    res.mars2,
    res.mars3,
    res.ad,
    res.bag,
    res.rf,
    res.boost,
    res.gblup
  )

colnames(result) <-
  c("R².Trein",
    "R².Val",
    "REQM.Trein",
    "REQM.Val",
    "n.fold",
    "method",
    "variable")

result <- result %>%
  mutate(
    variable = as.numeric(variable),
    method = as.factor(method),
    n.fold = as.factor(str_replace(n.fold, c("-", " "), "")),
    herd = as.factor(
      case_when(
        variable <= 6 ~ "30 %",
        variable > 6 & variable < 13 ~ "50 %",
        variable >= 13 ~ "80 %"
      )
    ),
    ngenes = as.factor(
      case_when(
        variable == 1 | variable == 7 | variable == 13 ~ 8,
        variable == 2 | variable == 8 | variable == 14 ~ 40,
        variable == 3 | variable == 9 | variable == 15 ~ 80,
        variable == 4 | variable == 10 | variable == 16 ~ 120,
        variable == 5 | variable == 11 | variable == 17 ~ 240,
        variable == 6 | variable == 12 | variable == 18 ~ 480
      )
    )
  )

barras<-result %>%
  group_by(ngenes,herd, method)%>%
  filter(n.fold != "Mean" & n.fold != "SD") %>%
  summarise(R2.mean = mean(`R².Val`)*100,
            REQM.mean = mean(REQM.Val)*100,
            R2.sd = sd(`R².Val`)*100,
            REQM.sd = sd(REQM.Val)*100)%>%
  ungroup()

barras<-barras %>%
  mutate(
    familia = as.factor(case_when(
      method == "MARS L" | method == "MARS Q" | method == "MARS C" ~ "Mars",
      method == "DT" | method == "RF" | method == "BAG" | method == "Boost" ~ "Tree",
      method == "G-BLUP" ~ "G-BLUP")
  ))

library(tidytext)

h_line <- barras %>%
  group_by(ngenes,herd)%>%
  summarise(mediar2 = mean(R2.mean),
            mediareqm = mean(REQM.mean))
  
levels(barras$method)<- c("BA",    "BO",  "DT",     "G-BLUP", "MARS 1", "MARS 2", "MARS 3",    "RF")
```

### Ploting results

#### Acuracy selective

```{r}
library(ggthemes) # Load

barras%>%
  mutate( varnames = fct_reorder2(method, R2.mean,familia))%>%
  ggplot(aes(x=varnames, y=R2.mean, fill=familia)) + 
  geom_col(alpha = 0.8, width = 0.85)+
  geom_point(colour = "#FC4E07")+
  geom_errorbar(aes(ymin=R2.mean-R2.sd, ymax=R2.mean+R2.sd), width=.2,
                position=position_dodge(.9))+
  scale_x_discrete(expand = expansion(add = 1))+
  scale_y_continuous(limits = c(0,100), expand = c(0,0),breaks = scales::breaks_width(10))+
  geom_hline(data = h_line,aes(yintercept = mediar2), linetype = "dashed", colour = "red")+
  facet_grid(herd~ngenes)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 60, hjust=1),
        text = element_text(size = 12), 
        panel.spacing = unit(1, "lines"),
        legend.position = "top")+
 labs(y = "R? (%)",x="", fill =  "")+
  scale_fill_gdocs()


ggsave(filename="r2.tiff", units = "cm", width = 30, height = 20, dpi = 300)

```
#### Acuracy predtive

```{r}
barras%>%
  mutate( varnames = fct_reorder2(method, REQM.mean,familia))%>%
  ggplot(aes(x=varnames, y=REQM.mean, fill=familia)) + 
  geom_col(alpha = 0.8, width = 0.85)+
  geom_point(colour = "#FC4E07")+
  geom_errorbar(aes(ymin=REQM.mean-REQM.sd, ymax=REQM.mean+REQM.sd), width=.2,
                position=position_dodge(.9))+
  scale_x_discrete(expand = expansion(add = 1))+
  geom_hline(data = h_line,aes(yintercept = mediareqm), linetype = "dashed", colour = "red")+
  facet_grid(herd~ngenes, scales = "free")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 60, hjust=1),
        text = element_text(size = 12), 
        panel.spacing = unit(1, "lines"),
        legend.position = "top")+
  labs(y = "RMSE",x="", fill =  "")+
  scale_fill_gdocs()

ggsave(filename="reqm.tiff", units = "cm", width = 30, height = 20, dpi = 300)
```

